src/audio/dp_queue.c

```diff
@@ -0,0 +1,351 @@
+#include <ipc/topology.h>
+
+LOG_MODULE_REGISTER(dp_queue, CONFIG_SOF_LOG_LEVEL);
+
+/* 393608d8-4188-11ee-be56-0242ac120002 */
+DECLARE_SOF_RT_UUID("dp_queue", dp_queue_uuid, 0x393608d8, 0x4188, 0x11ee,
+		 0xbe, 0x56, 0x02, 0x42, 0xac, 0x12, 0x20, 0x02);
+DECLARE_TR_CTX(dp_queue_tr, SOF_UUID(dp_queue_uuid), LOG_LEVEL_INFO);
+
+
+struct dp_queue {
+	struct sof_source source_api;  /**< src api handler */
+	struct sof_sink sink_api;      /**< sink api handler */
+
+	size_t ibs;
+	size_t obs;
+	uint32_t flags;		/* DP_QUEUE_MODE_* */
+
+	size_t data_buffer_size;
+	uint8_t __sparse_cache *data_buffer;
+
+	uint32_t write_offset;		/* to be modified by data producer ONLY */
+	uint32_t read_offset;		/* to be modified by data consumer ONLY */
+
+	bool hw_params_configured;
+	struct sof_audio_stream_params audio_stream_params;
+};
+
+static inline bool dp_queue_is_shared(struct dp_queue *dp_queue)
+{
+	return !!(dp_queue->flags & DP_QUEUE_MODE_SHARED);
+}
+
+static inline uint8_t __sparse_cache *dp_queue_buffer_end(struct dp_queue *dp_queue)
+{
+	return dp_queue->data_buffer + dp_queue->data_buffer_size;
+}
+
+static inline struct dp_queue *dp_queue_from_sink(struct sof_sink *sink)
+{
+	return container_of(sink, struct dp_queue, sink_api);
+}
+
+static inline struct dp_queue *dp_queue_from_source(struct sof_source *source)
+{
+	return container_of(source, struct dp_queue, source_api);
+}
+
+static inline void dp_queue_invalidate_shared(struct dp_queue *dp_queue,
+					      void __sparse_cache *ptr, size_t size)
+{
+	/* no cache required in case of not shared queue */
+	if (!dp_queue_is_shared(dp_queue))
+		return;
+
+	/* wrap-around? */
+	if ((uintptr_t)ptr + size > (uintptr_t)dp_queue_buffer_end(dp_queue)) {
+		/* writeback till the end of circular buffer */
+		dcache_invalidate_region
+			(ptr, (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr);
+		size -= (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr;
+		ptr = (__sparse_force void __sparse_cache *)dp_queue->data_buffer;
+	}
+	/* invalidate rest of data */
+	dcache_invalidate_region(ptr, size);
+}
+
+static inline void dp_queue_writeback_shared(struct dp_queue *dp_queue,
+					     void __sparse_cache *ptr, size_t size)
+{
+	/* no cache required in case of not shared queue */
+	if (!dp_queue_is_shared(dp_queue))
+		return;
+
+	/* wrap-around? */
+	if ((uintptr_t)ptr + size > (uintptr_t)dp_queue_buffer_end(dp_queue)) {
+		/* writeback till the end of circular buffer */
+		dcache_writeback_region
+			(ptr, (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr);
+		size -= (uintptr_t)dp_queue_buffer_end(dp_queue) - (uintptr_t)ptr;
+		ptr = (__sparse_force void __sparse_cache *)dp_queue->data_buffer;
+	}
+	/* writeback rest of data */
+	dcache_writeback_region(ptr, size);
+}
+
+struct sof_sink *dp_queue_get_sink(struct dp_queue *dp_queue)
+{
+	return &dp_queue->sink_api;
+}
+
+struct sof_source *dp_queue_get_source(struct dp_queue *dp_queue)
+{
+	return &dp_queue->source_api;
+}
```

Comment:
Not unless whole struct dp_queue is exported too
I think that it is worth some discussion - how to make something really private without harming performance...

Reply:
I'm certainly also for limiting the scope of types, but unfortunately that means reduced performance because of inability to use inline functions. If we were programming Firefox or LibreOffice or Gnome, I'd sacrifice performance for the sake of correctness. But for DSP firmware maybe we should do the opposite? Especially for code, used on hot paths during audio data processing

Reply:
a discussion opened https://github.com/thesofproject/sof/issues/8100